mlir/lib/Gradient/Transforms/GradMethods/Adjoint.cpp

```diff
@@ -72,31 +72,26 @@ void AdjointLowering::rewrite(GradOp op, PatternRewriter &rewriter) const
 func::FuncOp AdjointLowering::discardAndReturnReg(PatternRewriter &rewriter, Location loc,
                                                   func::FuncOp callee)
 {
-
-    std::vector<quantum::DeallocOp> deallocs;
-    callee.walk([&](quantum::DeallocOp dealloc) { deallocs.push_back(dealloc); });
+    SmallVector<quantum::DeallocOp> deallocs;
+    for (auto op : callee.getOps<quantum::DeallocOp>()) {
+        deallocs.push_back(op);
+    };
 
     // If there are no deallocs leave early then this transformation
     // is invalid. This is because the caller will expect a quantum register
     // as a return value.
     // Also, let's handle the simple case that is guaranteed at the moment.
-    bool invalid_transformation = deallocs.size() != 1;
-    if (invalid_transformation) {
-        callee.emitOpError() << "Invalid number of quantum registers: " << deallocs.size();
+    size_t numDeallocs = deallocs.size();
+    if (numDeallocs != 1) {
+        callee.emitOpError() << "Invalid number of quantum registers: " << numDeallocs;
         return callee;
     }
```

Comment:
@erick-xanadu While looking at this I was wondering whether this is a sufficient way to handle a failure inside of a rewrite pattern, because we then go on to return the callee which, if the pass execution continues, would be invalid I would think.

From within passes one can call `signalPassFailure()`, but I'm not sure about the best way to handle this inside a `Pattern::rewrite` method 🤔

Reply:
I'm not sure yet either, but perhaps:

```c
llvm::report_fatal_error
```

Reply:
@pengmai Do you know what the best way to signal an error from a rewrite pattern would be?

Reply:
In general, I'd consider switching the method from `rewrite` to `matchAndRewrite`. We could then return a `failure()` after the op error. This may require further refactoring of the callers to expect `LogicalResult`s as necessary.

If it's truly a fatal error indicating the user's program is completely invalid, crashing may be appropriate, though it's a bit tough to see where this would be a desirable user experience.

Reply:
The problem for this pass is if the pattern application fails then we can't compile the program, but it would also indicate that there is something wrong with the input IR (at the very least it's not a state that we currently ever expect).

Reply:
I think `report_fatal_error` makes sense here. I see the error more as guarding us against making changes which will violate the constraint of having more than one deallocOp. Sure, a user could write their own program using MLIR instead of the python interface, but I would consider that unlikely at the moment. Even if that were to happen, the IR would be invalid as we only handle one register allocation per quantum device and no quantum device's execution can interleave another making it necessary that there should only be one deallocation.

Somewhat orthogonal to this, but I think restructuring the removal of deallocOps could make this a bit cleaner. 

Wondering now about inlining and multiple quantum functions though... I'll look a bit into that.

Reply:
Right, so if this failure case would indicate a bug in the frontend, then crashing could be appropriate here.

One alternative to consider would be adding this invariant in the `verify` (or perhaps `verifySymbolUses`) method on the grad op (the callee must have exactly one `quantum::DeallocOp`). This pass could then operate assuming that verification was successful. This could actually be done independently of crashing/asserting the same condition within the pass itself.

Reply:
Yes! Something similar, is that I imagine could work is to have a function attribute that marks this function as a "quantum function" (short hand for having access to a quantum register and perhaps other properties we would like, e.g., "quantum function but no deallocation"). What do you think about that?

Reply:
At a certain point, it may be worthwhile to define the quantum function as its own op, particularly if the invariants and properties we expect it to have grow in complexity w.r.t. the base `func::FuncOp`. I know the MLIR authors expected more people to define their own `FunctionOpInterface`-conforming ops rather than using `func::FuncOp` out-of-the-box (though, of course, lots of people do the latter).

It may be trickier to attach a custom verifier to `func::FuncOp`. I'm not sure how one would go about this without making upstream changes.

Reply:
Good discussion!

> Wondering now about inlining and multiple quantum functions though... I'll look a bit into that. EDIT: Grad lowering happens before inlining. And even if it were to happen after, the functions are not really called until after grad lowering.

I specifically didn't enable inlining for quantum operations to avoid any issues

Reply:
I'll move this discussion to a new issue for a separate PR, thanks for the feedback :)